src/sentry/uptime/detectors/url_extraction.py

```diff
@@ -0,0 +1,53 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+from urllib.parse import urlsplit
+
+from django.core.exceptions import ValidationError
+from django.core.validators import URLValidator, validate_ipv46_address
+from tldextract import TLDExtract
+
+if TYPE_CHECKING:
+    pass
+
+url_validator = URLValidator()
+
+# Configure TLDExtract to not make external calls and just use a local snapshot of tld data.
+# Since this is instantiated at import time we don't want to make any external calls as
+# this could result in slow startup time or an error that prevents the server from starting.
+# We also don't want to be fetching this list from an unknown external source.
+extractor = TLDExtract(
+    cache_dir=None,
+    suffix_list_urls=(),
+    fallback_to_snapshot=True,
+)
+
+
+def extract_base_url(url: str | None) -> str | None:
+    """
+    Examines a url string and returns a url containing just the
+    protocol and fully qualified domain name from it.
+    Filters out invalid TLDs and domains with ip addresses. The
+    goal is to reduce the set of hostnames down so that we only test
+    hostnames that might actually return a valid response.
```

Comment:
Might be useful to toss in some examples. I don't fully understand what the extractor does.

Does it canonicalize URLs like `sentry.sentry.io` to just `sentry.io`?

Reply:
It takes a url like `http://sentry.io/hello` and returns the protocol + fqdn, which is `http://sentry.io`. So `http://sentry.sentry.io/issues` would be `http://sentry.sentry.io`

Reply:
ah ok, so it isn't doing anything with https://publicsuffix.org/ yet

Reply:
It will validate whether the tld is valid as part of it, but we're not doing anything to extract specific hostnames yet.